jina/drivers/evaluate.py

```diff
@@ -65,3 +66,56 @@ def _apply_all(self,
             evaluation.value = self.exec_fn(matches_ids, groundtruth_ids)
             evaluation.op_name = f'{self.id}-{self.exec.metric_name}'
             evaluation.ref_id = groundtruth.id
+
+
+class EncodeEvaluationDriver(BaseEvaluationDriver):
+    """Drivers used to pass `embedding` from documents and groundtruths to an executor and add the evaluation value
+    """
+
+    def __init__(self,
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _apply_all(self,
+                   groundtruth_pairs: Iterable[DocGroundtruthPair],
+                   *args,
+                   **kwargs) -> None:
+        for doc_groundtruth in groundtruth_pairs:
+            doc = doc_groundtruth.doc
+            groundtruth = doc_groundtruth.groundtruth
+            evaluation = doc.evaluations.add()
+            evaluation.value = self.exec_fn(pb2array(doc.embedding), pb2array(groundtruth.embedding))
+            evaluation.op_name = f'{self.id}-{self.exec.metric_name}'
+            evaluation.ref_id = groundtruth.id
+
+
+class CraftEvaluationDriver(BaseEvaluationDriver):
+    """Drivers used to pass a content field from documents and groundtruths to an executor and add the evaluation value
+    """
+
+    def __init__(self,
+                 field: str = 'text',
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.field = field
+
+    def _apply_all(self,
+                   groundtruth_pairs: Iterable[DocGroundtruthPair],
+                   *args,
+                   **kwargs) -> None:
+        for doc_groundtruth in groundtruth_pairs:
+            doc = doc_groundtruth.doc
+            groundtruth = doc_groundtruth.groundtruth
+            evaluation = doc.evaluations.add()
+
+            doc_content = getattr(doc, self.field)
+            gt_content = getattr(groundtruth, self.field)
+            if isinstance(doc_content, jina_pb2.NdArray):
+                doc_content = pb2array(doc_content)
+            if isinstance(gt_content, jina_pb2.NdArray):
+                gt_content = pb2array(gt_content)
+            evaluation.value = self.exec_fn(doc_content, gt_content)
```

Comment:
how can we handle the situation if one of the input is not `NdArray`? (I suppose we'll enocounter this situation because you did the check)

Reply:
This is something that it happens all over jina, you will have to provide an `executor` that accepts the input the `driver` is providing. (I would like to have an early assert to check which drivers and executors are compatible, but it has been decided otherwise).

At this point, the driver just assumes that the `executor` has the function signature that accepts the provided arguments (See all drivers have a call to `self.exec_fn` without having any constraint on the kind of signature it contains. So there should be a family of `executors` accepting `buffer` another for `blob` and another for `text`.

I think you have a really good point but it should be considered in general for all jina drivers and in another scope.

Reply:
as Python does not do type check, what @bwanglzu proposed makes sense but has low priority. No need to bother adding type check now